src/plots/cartesian/axes.js

```diff
@@ -82,21 +82,92 @@ axes.coerceRef = function(containerIn, containerOut, gd, attr, dflt, extraOption
     var refAttr = attr + 'ref';
     var attrDef = {};
 
-    if(!dflt) dflt = axlist[0] || extraOption;
+    if(!dflt) {
+        dflt = axlist[0] || (
+        ((extraOption === undefined) || (typeof(extraOption) === 'string')) ?
+            extraOption :
+            extraOption[0]
+        );
+    }
+    // so in the end, if dflt, axlist[], and extraOption were undefined, they
+    // are still undefined
     if(!extraOption) extraOption = dflt;
 
     // data-ref annotations are not supported in gl2d yet
 
+    var extraOptionAsList = extraOption;
+    if(extraOption) {
+        if(typeof(extraOption) === 'string') {
+            extraOptionAsList = [extraOption];
+        }
+    }
+
     attrDef[refAttr] = {
         valType: 'enumerated',
-        values: axlist.concat(extraOption ? [extraOption] : []),
+        values: axlist.concat(extraOptionAsList ? extraOptionAsList : []),
         dflt: dflt
     };
 
     // xref, yref
     return Lib.coerce(containerIn, containerOut, attrDef, refAttr);
 };
 
+/*
+ * An axis reference (e.g., the contents at the 'xref' key of an object) might
+ * have extra information appended. Extract the axis reference only.
+ *
+ * ar: the axis reference string
+ *
+ */
+axes.extractAxisFromAxisRef = function(ar) {
+    // TODO extractAxisFromAxisRef is more generic than this, this could be made
+    // more generic too
+    var mtch = (((ar !== undefined) && (typeof(ar) === 'string')) ?
+                ar.match(/^([xyz][0-9]*) domain/) :
+                undefined);
+    if(mtch) { return mtch[1]; }
+    return ar;
+};
+
+/*
+/*
+ * An axis reference (e.g., the contents at the 'xref' key of an object) might
+ * have extra information appended. Extract the extra information.
+ *
+ * ar: the axis reference string
+ *
+ */
+axes.extractInfoFromAxisRef = function(ar) {
+    if((ar !== undefined) && (typeof(ar) === 'string')) {
+        return ar.split(' ').slice(1).join(' ');
+    }
+    return ar;
+};
+
+/*
+ * Add the specified axis letter and number + " domain" to the extras for
+ * coerceRef.
+ *
+ * container: the object holding the [xyz]ref keys, e.g., a shape.
+ * axLetter: the letter of the axis of interest.
+ * coerceRefExtras: the current list of extras for coerceRef that will be
+ * appended to.
+ *
+ */
+axes.addAxRefDomainCoerceRefExtra = function(container, axLetter, coerceRefExtras) {
+    var axNumMatch = (
+            container[axLetter + 'ref'] ?
+            container[axLetter + 'ref'].match(/[xyz]([0-9]*)/) :
+            undefined
+        );
+    if(axNumMatch) {
+        var axNum = axNumMatch[1];
+        coerceRefExtras = coerceRefExtras.concat(
+                    (axNum !== undefined) ? [axLetter + axNum + ' domain'] : []);
```

Comment:
This approach doesn't tell you that `axLetter + axNum` is a valid axis - you could supply `xref = 'x99999 domain'` and it would be accepted.

`axes.coerceRef` is sufficiently specialized, I'd just give it another arg (maybe even just a boolean) to allow domain refs. Then after making `axlist` you can just duplicate each item plus `' domain'`.

Reply:
:+1: